Memory allocation

Manuzor opened the pull request:
A limited set of basic memory allocation mechanisms.
- [x] Stack memory type that manages an externally provided memory region.
  - [x] Alignment
- [x] Stack memory type with a given static size.
  - [x] Alignemnt
- [x] Heap memory type that provides a malloc-style implementation
  - [x] Alignemnt
  - [x] Merge adjacent free blocks when `Deallocate` is called.
    - Only considers blocks to the right from the deallocated one (see below).
  - ~~Possibility to navigate blocks bidirectionally.~~
    - ~~Improve the merging algorithm to also consider the blocks on the left.~~
  - [x] Reduce fragmentation issues with proper merging of adjacent free blocks.
    - This is done when finding a new free block during allocation. No need for a boundary tag at the end of each block.
- [x] System memory type that allocates memory by using system calls
  - [x] Alignemnt
- [x] Hybrid memory type that allocates from a primary memory type until it is out of memory, and then allocates from a secondary memory type
  - e.g. combine a `HeapAllocator` with the `SystemAllocator`.
- [x] An allocator type that wraps any memory type.
  - [x] New single object (constructed)
  - [x] Delete single object (destructed)
  - [x] New array (unconstructed)
  - [x] Delete array (undestructed)

Manuzor committed:
Add math functions IsOdd IsEven IsPowerOfTwo

Manuzor committed:
Remove krepel.span and replace by D slices

auto S = MakeSpan(10, SomePtr);
=>
auto S = SomePtr[0 .. 10];

Manuzor committed:
StackMemory and generic Allocator structs.

StackMemory can use static and custom memory blocks, or both.

Tests included as well.

Manuzor committed:
Working dynamic array implementation!

Manuzor committed:
Small test for Construct and Destruct.

Manuzor committed:
Add krepel.memory.allocation.SystemMemory.

This one uses system specific memory allocation procedures.

Manuzor committed:
Test HybridAllocator.

Manuzor committed:
:lipstick: HybridMemory.Deallocate made prettier.

Manuzor committed:
Implement working HeapAllocator!

It's not a perfect implementation, but it should work for now.

Manuzor committed:
Clean up heap memory allocation code.

Manuzor committed:
:lipstick: A bit of code cleanup.

Manuzor committed:
Fix heap block data structure.

It's now guaranteed to be size_t.sizeof.

Manuzor committed:
Fixes and documentation to different things in memory/allocation.d

Manuzor committed:
Merge branch 'master' into memory-allocation

Manuzor committed:
Implement alignment for system memory allocation.

Manuzor committed:
Fix overcomplicated heap memory allocation

pampersrocker commented:
Is there a specific reason to write all those containers as structs?

Manuzor committed:
Implement merging of free adjacent blocks when allocating.

Deallocation is now basically free (still requires a write to the given block to
mark it as deallocated, though). When finding a free block, we now merge all
adjacent free blocks we can find on the way, until we find one that is suitably
large. No need for a boundary tag at the end of a block.

Also add 0xDeadBeef to the end of each block :smile:

Manuzor commented:
What else? You can't allocate classes on the stack without some help of phobos template magic. So yes, the reason is because D doesn't let you allocate classes just like that. I think we could allocate classes using the allocators above, but I haven't tested them.

Classes in D are reference types.

``` d
class Foo {}
Foo Bar;
assert(Bar is null);
Bar = new Foo();
assert(Bar !is null);
```

Structs on the other hand work like in C++, i.e. they are value types.

I'm not sure but I think I prefer pointer-to-struct rather than a class.

``` d
struct Foo {}
Foo* Bar;
assert(Bar is null);
Bar = new Foo();
assert(Bas !is null);
```

Classes are originally meant to be used in conjunction with the garbage collector, of course. Sticking to structs only also has the advantage of simplifying many template cases. The biggest drawback is probably not having built-in virtual function support...

Manuzor commented:
I think I'm done with this for now. @pampersrocker please have a look at it.

Manuzor committed:
Class support for allocation.d and construction.d

pampersrocker commented on code/krepel/container/dynamic_array.d:
```diff
@@ -0,0 +1,148 @@
+    return Data[Index];
+  }
+
+  inout(ElementType)[] opIndex(T)(T[] Slice) inout
```
Whats this operator good for?

Manuzor replied:
It's the counter-part to `opSlice` which enables you to write something like this:

``` d
auto SomeSlice = MyArray[0..2]; // Get the first two elements.
// Will roughly translate to:
// auto SomeSlice = MyArray.opIndex(MyArray.opSlice(0, 2));
```

pampersrocker commented on code/krepel/container/dynamic_array.d:
```diff
@@ -0,0 +1,148 @@
+    AllocatorType* AllocatorPtr;
+  }
+
+  ElementType[] Capacity;
```
Naming misled me first into assuming this is an actual number, maybe something like `MemoryBlock` or `AllocatedMemory` ?

Manuzor replied:
I agree. The dynamic array implementation was just quickly hacked together to have something to use an allocator with :)

pampersrocker commented on code/krepel/memory/allocation.d:
```diff
@@ -0,0 +1,691 @@
+  /// Free the memory occupied by Instance without destructing it.
+  /// Note: If the memory type used does not support deallocation
+  ///       (e.g. StackMemory), this function does nothing.
+  void DeleteUndestructed(T)(T* Instance)
```
Undestructed implies for me destructing and then delete.
It has potential to be misunderstood, maybe `DeleteRaw` ?
Same for `New` `NewRaw` ?

Manuzor replied:
Well, it's not really raw for me... Raw implies you get a raw block of memory, but the returned block is already typed as `T`...

But why does _undestructed_ imply _destructing_ for you? It's quite verbose and I thought it would convey the idea of what it's doing... :disappointed:

pampersrocker replied:
Can be left as it is, I cannot come up with a better name for that. The documentation makes is clear.

pampersrocker commented on code/krepel/memory/allocation.d:
```diff
@@ -0,0 +1,691 @@
+    {
+      assert(IsInitialized, "Heap memory is not initialized.");
+      assert(Alignment == 0 || Alignment < ubyte.max,
+             "Alignment value is supposed to fit into 1 byte.");
```
Especially on consoles the alignment sizes can be up to 2MB! (Not necessary for now)

Manuzor replied:
Yea, this can be temporary for now. In the current implementation I just expect there to be only smaller alignment values. Maybe it's not correct to do this check for the alignment, but for the actual padding value instead, since that's the value I'm storing in that 1 byte.

pampersrocker commented on code/krepel/meta.d:
```diff
@@ -109,12 +102,6 @@ template AnySatisfy(alias Func, Args...)
   debug(krepel_meta_AnySatisy) pragma(msg, "[krepel_meta_AnySatisy] Result = ", AnySatisfy.stringof);
```
AnySatis**f**y

Manuzor replied:
At least it's consistent... :sweat_smile:

Manuzor committed:
Fix typo satisy => satisfy

Manuzor committed:
Rename Capacity => AvailableMemory in DynamicArray

Manuzor committed:
:lipstick: Improve code readability in Construct.

Manuzor committed:
Remove seemingly unnecessary overload of opIndex.

Even though I've read about opIndex(SliceResult) being necessary, the compiler
definitely calls opSlice directly.

Manuzor commented:
Any thoughts?

pampersrocker commented on code/krepel/memory/construction.d:
```diff
@@ -0,0 +1,172 @@
+    Element.Value = 1337;
+  }
+  DestructArray(Array);
+  foreach(ref Element; Array) assert(Element.Value == 42);
```
According to the destructor, shouldn't this be 666 ?

Manuzor replied:
Yea, I thought so as well... I've modelled the destruction logic after D's `destoy()` functions' behavior, which does not leave the object in an uninitialized state. It reverts it to the initial state `T.init`. I forgot to test that the destructor is actually called though.

pampersrocker commented on code/krepel/memory/allocation.d:
```diff
@@ -0,0 +1,725 @@
+  assert(Block3.length == 16);
+  assert(Heap1.Contains(Block1));
+  auto Block4 = Hybrid.Allocate(16);
+  assert(Block4);
```
No test for the actual hybrid stuff? You exactly stop, when trying to do allocate from the second block.

Manuzor committed:
Fix destructing structs.

Manuzor committed:
Don't blit initial data after destruction.

Also manually destruct all members of a non-class object.

Manuzor committed:
Complete HybridAllocator test.
